Add a landing section to stylebook tabs

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Closes #66517.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Open stylebook by clicking on Styles from the left hand side black sidebar, then clicking on Preview and selecting Style book from the dropdown.
2. See that the first blocks to appear are the same as in the landing page designs from #66517.
3. Open stylebook by clicking on Styles in the top right hand side of the editable canvas, and selecting Style Book from the top of the right hand side bar.
4. See there's now a Landing tab.
5. In either view, clicking on any of the blocks should take you to the relevant section of the global styles interface.

Given that the stylebook view in the left hand sidebar has everything in one long scroll, I deduped the items from the Landing section and added them in at the top.

## Screenshots or screencast <!-- if applicable -->

<img width="1455" alt="Screenshot 2024-10-29 at 1 57 12 pm" src="https://github.com/user-attachments/assets/79aae4ff-1df5-4ea6-88dc-391d370dda5a">


<img width="1425" alt="Screenshot 2024-11-06 at 10 00 36 am" src="https://github.com/user-attachments/assets/56cc0825-ba14-4e81-aa1d-8928ebc0b4f9">

tellthemachines committed:
Add a landing section to stylebook tabs

tellthemachines committed:
Check that theme palette exists

tellthemachines committed:
dedupe examples for single page stylebook

tellthemachines committed:
Fix Heading example resizing

tellthemachines committed:
Fix e2e tests

tellthemachines committed:
Fix e2e

ramonjd commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -203,6 +203,22 @@ function StyleBook( {
+		return (
+			example.category !== 'landing' &&
+			! landingCategoryExamples.examples.find(
+				( landingExample ) => landingExample.name === example.name
```
Nice work filtering out duplicates on the single page 💯

aaronrobertshaw replied:
Should we memoize the collection of these deduped single page examples?

ramonjd commented on packages/edit-site/src/components/style-book/constants.ts:
```diff
@@ -107,6 +107,11 @@ export const STYLE_BOOK_THEME_SUBCATEGORIES: Omit<
 export const STYLE_BOOK_CATEGORIES: StyleBookCategory[] = [
+	{
+		slug: 'landing',
+		title: __( 'Landing' ),
```
I know "Landing" has been bandied about, but what does it mean exactly to a user?

From https://github.com/WordPress/gutenberg/issues/53431:

> It serves as a quick glanceable "poster view" for the site style guide

I don't have a better alternative — 'Overview'? — but I suppose it could go in the direction of "Theme style guide" if and when we allow patterns and/or theme devs to add to it.

aaronrobertshaw replied:
Personally, I think I lean towards "Overview" rather than "Landing" as being more meaningful for more people. 

IIRC there was some talk recently about quirky things and personality being something missed lately. A left field option could be "At a glance". It's not a single word but is still relatively short. Once it's translated that could be a different story though.

andrewserong replied:
Great question. I'll just ping @richtabor for discussion on the name for the category as he often has good ideas for these things!

jasmussen replied:
Connecting some dots to my wall of [text here](https://github.com/WordPress/gutenberg/pull/66545#issuecomment-2459211973), I expect the tabs to eventually go away in favor of drilldown-based contextual navigation. Instead of clicking a tab, you click "Typography" in global styles, and the preview updates to match.

In that vein, the name of the tab stops becoming user visible.

ramonjd commented on packages/edit-site/src/components/style-book/examples.tsx:
```diff
@@ -62,6 +62,87 @@ function getColorExamples( colors: MultiOriginPalettes ): BlockExample[] {
+		title: __( 'Headings' ),
+		category: 'landing',
+		blocks: createBlock( 'core/heading', {
+			content: `AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz0123456789{(...)},?!*&:;_@#$`,
```
Wondering if this should be a translatable string? Mainly for the letter characters in case it's important to show off certain characters, e.g., umlauts, or use language-specific alphabets.

aaronrobertshaw replied:
+1 on the translation front. I think it definitely fits with the overall extensibility and customization goals for the Style Book.

ramonjd commented on packages/edit-site/src/components/style-book/examples.tsx:
```diff
@@ -62,6 +62,87 @@ function getColorExamples( colors: MultiOriginPalettes ): BlockExample[] {
+		title: __( 'Paragraph' ),
+		category: 'landing',
+		blocks: createBlock( 'core/paragraph', {
+			content: `There was an Old Man of Vienna, 
```
👍🏻 

Should we make this translatable too?

aaronrobertshaw replied:
+1

jasmussen replied:
Noting some thoughts [here](https://github.com/WordPress/gutenberg/pull/66545#issuecomment-2459211973), that are related.

ramonjd reviewed (commented):
This is fantastic straight out of the box, thanks for working on it.

I feel it's in a pretty good place to merge. Keen to hear what other folks think 

 I left a few minor questions that are not blockers.

aaronrobertshaw reviewed (commented):
Nice work @tellthemachines ✨ 

This tests very nicely for me as well.

My vote would be to make the small tweaks @ramonjd noted (e.g. naming & translation), merge, and then iterate from there. 

As an aside, it feels a bit odd to me having the color palettes split on the single page view of the Style Book. That's per the design at this point so not an issue for this PR, just something I hope comes out in the wash during future iterations.

andrewserong reviewed (commented):
Nice work here, this is looking really good to me, too!

~Not introduced in this PR I'm sure, but~ I noticed while testing that the break point for smaller sizes when reducing the resizable frame is a bit squished / doesn't fill out the horizontal space available:

<img width="1279" alt="image" src="https://github.com/user-attachments/assets/01c6eade-e9a7-4939-a1e3-f0d81f4feb4b">

Edit: looks like it was the change to the width rule. I've added a comment (https://github.com/WordPress/gutenberg/pull/66545#discussion_r1830492215) 🙂

andrewserong commented on packages/edit-site/src/components/style-book/constants.ts:
```diff
@@ -247,7 +252,7 @@ export const STYLE_BOOK_IFRAME_STYLES = `
 
 	.edit-site-style-book__example-preview {
-		width: 100%;
+		width: calc(100% - 120px);
```
It looks like this results in the reduced width in narrower sizes:

| Before this rule | With this rule |
| --- | --- |
| <img width="545" alt="image" src="https://github.com/user-attachments/assets/c02e31b7-2465-4b7a-8c97-7d445fd7102e"> | <img width="523" alt="image" src="https://github.com/user-attachments/assets/6ef9b8ab-d4b6-4929-849c-643705ab712b"> |

Would it be worth using a media query or other way to guard this?

tellthemachines replied:
Done!

jasmussen commented:
Nice, good progress!

![2024-11-06 10 32 18 localhost f3f82e50e994](https://github.com/user-attachments/assets/9797fb1e-f1e1-4daa-9b39-f7d0cee52b2e)

It's unclear to me how much we need to accomplish in one PR vs another, knowing that at the end of the effort we want to be as close to the following as possible:

![Pages](https://github.com/user-attachments/assets/97b5efe8-75eb-4fae-a54e-612778bac44b)

_The above is perhaps easier to see in Figma, [here](https://www.figma.com/design/guZBzJKVdWLcNeWCZ7Rvx1/Style-Book?node-id=10-3598&t=ttLX2GAJAxEM1Ki0-1)._

Focusing first on this landing page, it might just be nicer to get closer to those initial mockups visually. Mainly standing out is that the landing page is heavily curated, showing both Heading and Paragraph blocks in a column pattern, rather than just as lists of blocks. 

Precisely what we show as text there is a separate task on its own. The old man of Senna is a fun little blurb, but to better match this pattern, we'll want something longer. And since we don't support `text-columns`, it will likely need to be two paragraphs, one in each column of the pattern (and the columns block should stack on mobile).

![Screenshot 2024-11-06 at 10 59 23](https://github.com/user-attachments/assets/583e60b9-0cac-4c3b-85f1-6d255b77b727)

We _do_ want custom text on this landing page, and it should be translatable ideally. But we also want to get closer to the pattern. If not to start, then later, so at some point translations of this initial piece will become obsolete.

There's also a new treatment for the style book headings, featuring horizontal separators. Note also how the new style book designs appear to be leveraging wide and content width alignments, with everything on the landing page being _wide_, but many blocks and pieces on subpages being _content-width_. Example:

![Screenshot 2024-11-06 at 11 03 15](https://github.com/user-attachments/assets/eed986a4-07bf-4f07-bf0b-26a6fc99453f)

At some point, this landing page should be the default thing you see in the new style book location in Site View. As part of this effort we are replacing the tabs at the top, with contextual navigation of the new left sidebar. In this view, the behavior is like so:

* Clicking "Typography" shows the Typography tab contents.
* Clicking "Colors" shows the color tab contents.
* Clicking "Blocks" shows _all the other blocks_, though we need to figure out the categorization and how.
* Clicking any individual block, shows just that block, and with an opportunity to also show every state of this block. 

Visually going through the flow, you start here:

<img width="1840" alt="i2" src="https://github.com/user-attachments/assets/a9b0bdda-c6d4-4296-a54c-83779b2378f2">

Click Typography, and you see the contents of the typography tab (note: rough mockup!):

<img width="1840" alt="typography" src="https://github.com/user-attachments/assets/849a4ec3-cb2d-46ad-a521-1c042adf6948">

Click Colors, and you see the contents of the color tab (note: rough mockup!):

<img width="1840" alt="colors" src="https://github.com/user-attachments/assets/268d88a8-3cac-439c-b004-5dca448ce643">

Click Blocks → Navigation, and you see an exploded view of the navigation block (note: rough mockup!):

<img width="1840" alt="navigation block" src="https://github.com/user-attachments/assets/4a550b0c-6e28-4516-a4c2-2e3f43f630b5">

_The navigation block with states is [related to this](https://github.com/WordPress/gutenberg/issues/38277#issuecomment-2306449475)._ 

Where we place media and site identity, is as-yet unclear. So we may indeed want to include those on the landing page, for now. 

Separate to this PR, but related as far as tasks, there's an effort needed to update and improve the default placeholder content for many of our blocks. Referring to mont blanc here: 

<img width="343" alt="image" src="https://github.com/user-attachments/assets/76591268-dda1-424d-bb9a-f45794800acd">

Whether we can use the images @beafialho included in her mockup, we need to check the license. But we definitely should update these, as they are far better:

<img width="1092" alt="image" src="https://github.com/user-attachments/assets/f4a83be5-39e1-4519-9027-bc1b9bdd5eb4">

Some of these pieces would be good to get @beafialho advice on, in case there's any nuance that has changed since she created the initial mockups. But the main point is: the landing page is more like a block editor page with patterns, where all the other subsections show the full breadth of blocks available, even if they also have somewhat updated categories.

**In summary**

There are some immediate tasks to get closer to the landing page curation and layout as outlined, specifically this layout:
![Screenshot 2024-11-06 at 11 11 50](https://github.com/user-attachments/assets/5941f763-3b08-4773-b597-5256c332978c)

The heading and subheading style is one task, applying a heading + columns pattern is another. @beafialho not sure how busy you are, but would you be able to work with @tellthemachines to supply such a pattern? If you can build it in the editor, I imagine we can use it here. 

As far as the other tasks—updating the drilldown behavior to retire the tabs, showing an exploded view of block states in the Blocks drilldown—we can follow up with separately, and probably the task of updating default content (especially for Image and Gallery blocks) I or some other designers can do. 

Does that breakdown make sense? Thanks for working on this!

beafialho commented:
> Whether we can use the images @beafialho included in her mockup, we need to check the license. But we definitely should update these, as they are far better:

Thanks for the ping @joen, I double checked the image sources and just in case I have replaced the ones that had the "Free" label on RawPixel with similar ones that are Public Domain. That way, we can use them all. Here are the links if they happen to be necessary:

- https://www.rawpixel.com/image/5949471/free-public-domain-cc0-photo
- https://www.rawpixel.com/image/4031610/coastal-wolf
- https://stocksnap.io/photo/nature-green-66J4T150FP
- https://www.rawpixel.com/image/4039528/photo-image-background-nature-mountain
- https://www.rawpixel.com/image/434475/free-photo-image-eagle-mountain-animal-bird
- https://www.rawpixel.com/image/5923986/photo-image-public-domain-trees-woods

> The heading and subheading style is one task, applying a heading + columns pattern is another. @beafialho not sure how busy you are, but would you be able to work with @tellthemachines to supply such a pattern? If you can build it in the editor, I imagine we can use it here.

Just to clarify, do you mean building this pattern in the editor?

![Captura de ecrã 2024-11-06, às 14 58 14](https://github.com/user-attachments/assets/b520c13e-c481-4f16-b63f-0aa2f5bc4aa9)

jasmussen commented:
> Just to clarify, do you mean building this pattern in the editor?

Yes exactly, I think having this pattern would be useful. It's unclear whether we can/should use `text-columns` for that particular layout or not. It's also an opportunity to think about what phrase we want to include there, whether it's something we author, or not. I quite appreciate the Edward Lear limerick that is currently present in this PR (the old man of Vienna), which also happens to be public domain. It adds some character. At the same time, it's not long enough to fit the pattern you designed.

tellthemachines committed:
Fix styles on small breakpoint

tellthemachines commented:
> Just to clarify, do you mean building this pattern in the editor?

For the stylebook example I'll be creating the blocks dynamically with `createBlock()` so I don't need actual block markup. In this case, to include Headings and Paragraph in the same example (so they're under the same section heading), it could be a Group with nested Columns or it could even be a Grid.

tellthemachines committed:
Update landing text examples

tellthemachines committed:
Make two col layout responsive

tellthemachines committed:
Make translatable

tellthemachines commented:
Thanks for all the feedback and suggestions folks! I think I've addressed the major points now:

* Example block strings are translatable 

Note that the two Paragraph blocks might not make sense as self-contained units in other languages. Their length will probably also vary depending on the language. The problem is translators see strings out of context, so there's no way to advise them on making both those strings the same length. Given the text content isn't important here, we might want to consider leaving these strings untranslated so we can predict their size accurately.

Even if we were able to use CSS columns here instead of a 2-col Grid, we couldn't ensure both cols had the same amount of text.

* Examples are full width in small viewports
* Text content has been updated to match the design, including 2-col layout:

<img width="1477" alt="Screenshot 2024-11-07 at 4 57 43 pm" src="https://github.com/user-attachments/assets/5cc080f0-d2fd-41e8-9301-b0253884e3e4">

Let me know if I missed anything!

I reckon further design updates to the section headings, as well as creating individual sections for colors, typography etc. would be best handled as separate PRs.

jasmussen commented:
Looks good. Would it be good to extract [these notes](https://github.com/WordPress/gutenberg/pull/66545#issuecomment-2459211973) into one or several issues, so anyone can pick them up?

Thank you for updating the text, looks great. I guess we don't need your pattern after all, @beafialho! One thing we do need to figure out, and you mention this yourself: we need to thread the needle on exactly what text to show here. 

How about this text, at least for now, and then other PRs can update them if they have further improvements.

> All these hardships we face now are but the beginning of great adventures, Sancho. The greater the danger, the greater the glory. Life, my friend, is full of mountains to climb and dragons to slay, and though our armor may be battered and our strength may falter, it is only those who press on who can hope to claim the prize of their dreams. For the world is a field of challenges, a place where courage is not just tested but tempered, forged anew with each step into the unknown. Remember, Sancho, that courage is found not in the easy paths, but in those riddled with trials and tribulations.
> 
> So let us lift our lances high and keep our spirits even higher. Though we may stumble, though we may encounter enemies both seen and unseen, the road forward is one we must travel. And if the world sees our journey as madness, let them, for we know the true purpose of our quest. Each challenge makes us stronger, each trial shapes us into men of unbreakable spirit. And perhaps, one day, when we look back on these struggles, we shall see that these very hardships were the greatest gifts bestowed upon us, for they showed us who we truly are.

That would look about like this:

![Screenshot 2024-11-07 at 09 43 18](https://github.com/user-attachments/assets/11d303aa-67a6-4879-be1d-3aac4b4d0044)

It's not quite perfect, since the left paragraph is longer than the right one. You mentioned potentially using `text-columns` after all? Can we do that here? Otherwise, the two columns may be okay after all. On a meta note, it may be helpful that this is a celebrated and broadly translated public domain work, so for the international translations, there may be a nice and obvious source to copy from.

beafialho commented:
> How about this text, at least for now, and then other PRs can update them if they have further improvements.

The text you suggested @jasmussen also looks good to me, but just to offer some context, my initial idea was to use some informative/educational content about each block as demo text. I did that for the headings, post content, quote and pullquote blocks:

![Captura de ecrã 2024-11-07, às 10 35 58](https://github.com/user-attachments/assets/56f50106-d607-4cde-aa86-9fe6b0540503)

![Captura de ecrã 2024-11-07, às 10 36 07](https://github.com/user-attachments/assets/17dba581-8df3-4e54-8398-77330daac54f)

![Captura de ecrã 2024-11-07, às 10 36 14](https://github.com/user-attachments/assets/7f1d9ebc-e1bc-42e7-8b44-74ff44add2ea)

For headings, the full text would be:

> Headings provide hierarchical structure and organization to the content. They are used to denote different sections or sub-sections within a webpage, making it easier for users to navigate and comprehend the information presented.

jasmussen commented:
Ah, makes perfect sense. Though we still need the precise text you authored for those two paragraphs, though. And either it's _one_ paragraph split using `text-columns`, or it's two complete paragraphs. For the moment, it's two paragraph blocks, but one ends mid-sentence and the other starts mid-sentence.

beafialho commented:
Sure, I used the text below. If I'm understanding correctly, it's a problem that it starts mid-sentence?

In the design, that would be fine because as you increase or decrease text size, the paragraphs would normally redistribute across columns, like [this example](https://www.fontfabric.com/fonts/alkes/#font-tester).

> A paragraph in a website refers to a distinct block of text that is used to present and organize information. It is a fundamental unit of content in web design and is typically composed of a group of related sentences or thoughts focused on a particular topic or idea. Paragraphs play a crucial role in improving the readability and user experience of a website. They break down the text into smaller, manageable chunks, allowing readers to scan and comprehend the content more easily. Additionally, paragraphs help structure the flow of information and provide logical breaks between different concepts or pieces of information. In terms of formatting, paragraphs in websites are commonly denoted by a vertical gap or indentation between each block of text. This visual separation helps visually distinguish one paragraph from another, creating a clear and organized layout that guides the reader through the content smoothly.

https://github.com/user-attachments/assets/84dc0c82-c13f-4f99-8554-39ac52642170

jasmussen commented:
The problem is in the PR currently, it's these two "paragraphs":

> A paragraph in a website refers to a distinct block of text that is used to present and organize information. It is a fundamental unit of content in web design and is typically composed of a group of related sentences or thoughts focused on a particular topic or idea. Paragraphs play a crucial role in improving the readability and user experience of a website. They break down the

> text into smaller, manageable chunks, allowing readers to scan and comprehend the content more easily. Additionally, paragraphs help structure the flow of information and provide logical breaks between different concepts or pieces of information. In terms of formatting, paragraphs in websites are commonly denoted by a vertical gap or indentation between each block

An ideal solution would be [this](https://codepen.io/joen/pen/OJKBwKE?editors=1100), i.e. use `column-count: 2;` (sorry I previously referred to this as `text-columns`, my mistake).

Another solution would be to break the text at a proper period. Like so:

| Col 1 | Col 2 |
|--------|--------|
| A paragraph in a website refers to a distinct block of text that is used to present and organize information. It is a fundamental unit of content in web design and is typically composed of a group of related sentences or thoughts focused on a particular topic or idea. Paragraphs play a crucial role in improving the readability and user experience of a website. They break down the text into smaller, manageable chunks, allowing readers to scan and comprehend the content more easily. | Additionally, paragraphs help structure the flow of information and provide logical breaks between different concepts or pieces of information. In terms of formatting, paragraphs in websites are commonly denoted by a vertical gap or indentation between each block of text. This visual separation helps visually distinguish one paragraph from another, creating a clear and organized layout that guides the reader through the content smoothly. | 

 How do you feel about the latter?

beafialho commented:
> Another solution would be to break the text at a proper period. Like so:

> How do you feel about the latter?


I see! Yes, that sounds good to me.

tellthemachines commented:
> An ideal solution would be [this](https://codepen.io/joen/pen/OJKBwKE?editors=1100), i.e. use `column-count: 2;` (sorry I previously referred to this as `text-columns`, my mistake).

This isn't an existing feature in the block editor though, is it? Wouldn't it be better to stick to a presentation that is reproducible with blocks in the UI? Otherwise folks might think that they can have 2-col Paragraphs, when that isn't possible without some hacky extra CSS.

aaronrobertshaw commented:
> This isn't an existing feature in the block editor though, is it?

FWIW a while back text column (column-count) support was [added to typography block supports](https://github.com/WordPress/gutenberg/pull/33587). Efforts there weren't a high priority so it was put on the backburner and forgotten about.

There are still open issues to support it for blocks like [paragraph](https://github.com/WordPress/gutenberg/issues/27118) and [post-excerpt](https://github.com/WordPress/gutenberg/issues/25459). 

If I'm not forgetting any blockers, I don't think it would take a lot to enable that support. The question is if we really want that even as an optional (hidden by default) typography control on the paragraph block.

tellthemachines commented:
> FWIW a while back text column (column-count) support was [added to typography block supports](https://github.com/WordPress/gutenberg/pull/33587). Efforts there weren't a high priority so it was put on the backburner and forgotten about.

I had no idea this existed 😮 

I guess we could consider adding support to the Paragraph block (was there any reason not to do so initially?), but I'm not sure how useful it'll be without a wide alignment control because with a standard content width two columns look a little squished:

<img width="698" alt="Screenshot 2024-11-11 at 11 33 52 am" src="https://github.com/user-attachments/assets/7d0a5ddb-b31c-40a6-bea0-95ba86cc6d95">


For now, I'll go with updating the text to the two separate paragraphs mentioned above and we'll see how that works.

tellthemachines committed:
Change "landing" to "overview"

tellthemachines commented:
Ok I've updated the Paragraphs and I've also changed the section name to "Overview" as that seems to be preferred.

One thing remaining to do: link the new "Text" section in overview to a section in global styles. I'm guessing that would be the global typography panel? Currently it doesn't link anywhere:

<img width="1690" alt="Screenshot 2024-11-11 at 12 03 01 pm" src="https://github.com/user-attachments/assets/b7af9fb8-c0fc-4bfd-b034-eec60a560009">

tellthemachines committed:
Add link to typography section and update e2e

tellthemachines commented:
I've pushed an update that changes the name of the "Text" section to "Typography" to better match the corresponding global styles section, and linked it to typography in global styles.

This should now be ready for a final round of testing and reviews!

ramonjd commented on packages/edit-site/src/components/sidebar-global-styles-wrapper/index.js:
```diff
@@ -131,6 +131,10 @@ export default function GlobalStylesUIWrapper() {
 							// Go to color palettes Global Styles.
 							onPathChange( '/colors/palette' );
 							return;
+						} else if ( blockName === 'typography' ) {
+							// Go to typography Global Styles.
+							onPathChange( '/typography' );
+							return;
 						}
```
```suggestion
						}

						if ( blockName === 'typography' ) {
							// Go to typography Global Styles.
							onPathChange( '/typography' );
							return;
						}
```

I think the `else if` is redundant due to the previous early return?

Also it looks like the `/typography` route handling needs to happen in the editor too: https://github.com/WordPress/gutenberg/blob/2a18aae3768da80f94e437c16e8d56e5a9a45e03/packages/edit-site/src/components/global-styles/ui.js#L194-L195

tellthemachines replied:
Fixed, thanks for spotting that!

ramonjd reviewed (approved):
🚢 

This is working well for me in both view and edit modes of the site editor. Thank you!!

Tested in Firefox, Chrome, Safari and Edge. 

There's just one outstanding update to add the `/typography` route to editor style book `onSelect` callback.

tellthemachines committed:
Add typography path in editor canvas stylebook.
